### PR TITLE
BP-2094 - refactor data builders and availability validators pool

### PR DESCRIPTION
**Pull request by LucianTuriacArnia**

### CODE REVIEW

File: Gateway/Request/BasicParameter/CurrencyDataBuilder.php [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+use Buckaroo\Magento2\Exception;
+use Buckaroo\Magento2\Gateway\Helper\SubjectReader;
+use Buckaroo\Magento2\Model\ConfigProvider\Method\Factory;
+use Magento\Payment\Gateway\Request\BuilderInterface;
+use Magento\Sales\Model\Order;
+
+class CurrencyDataBuilder implements BuilderInterface
+{
+    private const CURRENCY = 'currency';
```
Comment (harli91) [inline Gateway/Request/BasicParameter/CurrencyDataBuilder.php:13]: maybe rename this to CURRENCY_KEY ?

File: Gateway/Request/BasicParameter/CurrencyDataBuilder.php [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    /**
+     * Set Allowed Currencies
+     *
+     * @param \Buckaroo\Magento2\Model\Method\BuckarooAdapter $methodInstance
+     * @return $this
+     * @throws \Buckaroo\Magento2\Exception
+     */
+    private function setAllowedCurrencies($methodInstance)
+    {
+        $method = $methodInstance->getCode() ?? 'buckaroo_magento2_ideal';
```
Comment (harli91) [inline Gateway/Request/BasicParameter/CurrencyDataBuilder.php:118]: why set here the default as `buckaroo_magento2_ideal`  and not throw a exception?

File: Gateway/Response/CaptureHandler.php [hunk 3]
```diff
@@ -24,13 +24,15 @@ class CaptureHandler extends AbstractResponseHandler implements HandlerInterface
...
         ) {
             throw new \InvalidArgumentException('Payment data object should be provided');
         }
 
-        if (!isset($response['object'])
+        if (
+            !isset($response['object'])
             || !$response['object'] instanceof TransactionResponse
         ) {
             throw new \InvalidArgumentException('Data must be an instance of "TransactionResponse"');
```
Comment (harli91) [inline Gateway/Response/CaptureHandler.php:38]: these checks are in multiple files, maybe refactor them into a abstract class/ trait

Comment (LucianTuriacArnia) [reply Gateway/Response/CaptureHandler.php:38]: Thank you, I created this class Buckaroo\Magento2\Gateway\Helper\SubjectReader. I will update the code to be used

File: Gateway/Request/BasicParameter/CurrencyDataBuilder.php (hunk 2 above)
Comment (LucianTuriacArnia) [reply Gateway/Request/BasicParameter/CurrencyDataBuilder.php:118]: app/code/Buckaroo/Magento2/Gateway/Http/TransactionBuilder/Order.php:353 I copied the logic from here, this file was modified in commit: BP-391 Add payment method PayPerEmail (Create Paylink only)  9/23/2020 11:42 AM

**Review by harli91 — APPROVED** [review]

**Pull request merged**